text/0000-prelude-2021.md

```diff
@@ -0,0 +1,164 @@
+- Feature Name: `prelude_2021`
+- Start Date: 2021-02-16
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+A new prelude for the 2021 edition, featuring several extra traits.
+
+# Motivation
+[motivation]: #motivation
+
+While types and free functions can be added to the prelude independent of edition boundaries, the same is not true for traits. Adding a trait to the prelude can cause compatibility issues because calls to methods named the same as methods of the newly in scope traits can become ambiguous. Because editions are opt-in, they represent an opportunity to add new traits to the prelude.
+
+The expected outcome is to make the newly added traits more accessible to users, potentially stimulating their use compared to alternatives.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The compiler currently brings all items from `std::prelude::v1` into scope for each module (for code using `std`). Crates that declare usage of the 2021 edition will instead get items from `std::prelude::edition2021` imported into scope.
+
+This RFC proposes to add the following traits to `std::prelude::edition2021`, but not `std::prelude::v1`:
+
+- `TryFrom`/`TryInto`
+- `FromIterator`
+
+Apart from the 2021-specific prelude, this RFC also proposes to add several items to preludes for all editions. These changes do not have to be aligned with an edition boundary, since there is no compatibility hazard (explicitly imported names always shadow prelude names).
+
+- `std::{mem, fmt}`
```

Comment:
I feel like modules are a pretty large change here, so far we've only ever had types, traits, and functions in the prelude, if modules are included then people may think they are _crates_. This drawback should be called out and addressed.

Reply:
Given that e2018 makes all cratenames implicitly present in every scope, I personally am *very* skeptical of having any other source of implicit inclusion of lowercase symbols.

Currently, the preludes only import one function: `core::mem::drop`. All other lowercase symbols imported by the preludes are macros, which can only be used with `macname!` or `#[macname]`, and are not able to visually collide with crate, module, or function names in source. I *tentatively* would not disagree with placing additional functions in the prelude, but I *firmly* believe that placing module names in the prelude is a mistake for exactly the reason you outline.

Reply:
Should the RFC specify what happens if the module name collides with a crate? There are already `mem` and `fmt` crates.

Reply:
No, the prelude is just a glob import, glob imports have well-defined behavior around this.

Reply:
@mbartlett21 it's also possible that other packages explicitly specify the cargo `lib.name` to be `mem` or `fmt`. Not sure if there are any of those, but it's quite likely that this happens in private (as in not released on crates.io) monorepo projects. For example, when I write a binary that uses some ad-hoc procedural macros, I would create a `my-codegen` package in the same workspace with `lib.name` as a very short word, e.g. `mac` or `gen`. I am unsure whether other people would name their crates `mem` or `fmt` in a similar manner.

Reply:
While adding mem/fmt to the prelude could be pragmatic, I think the drawback of introducing a "separate root" (apparently, effectively) like `mem` makes the whole thing harder to understand for users. I'd argue for lifting these functions to `std::swap` etc (as mentioned elsewhere) or other solutions, like no change.